plugins/woocommerce-admin/client/analytics/components/report-chart/index.js

```diff
@@ -91,38 +97,100 @@ export class ReportChart extends Component {
 			query,
 			defaultDateRange
 		);
-		const chartData = primaryData.data.intervals.map( function (
-			interval,
-			index
-		) {
-			const secondaryDate = getPreviousDate(
+
+		const primarydataContainsLeapYear = dataContainsLeapYear( primaryData );
+		const secondarydataContainsLeapYear =
+			dataContainsLeapYear( secondaryData );
+		const primaryDataIntervals = [ ...primaryData.data.intervals ];
+		const secondaryDataIntervals = [ ...secondaryData.data.intervals ];
+
+		const chartData = [];
+
+		for ( let index = 0; index < primaryDataIntervals.length; index++ ) {
+			const interval = primaryDataIntervals[ index ];
+
+			const secondaryDateMoment = getPreviousDate(
 				interval.date_start,
 				primary.after,
 				secondary.after,
 				query.compare,
 				currentInterval
 			);
 
-			const secondaryInterval = secondaryData.data.intervals[ index ];
-			return {
-				date: formatDate( 'Y-m-d\\TH:i:s', interval.date_start ),
+			const primaryDateFormatted = formatDate(
+				'Y-m-d\\TH:i:s',
+				interval.date_start
+			);
+			const primaryLabel = `${ primary.label } (${ primary.range })`;
+			const primaryLabelDate = interval.date_start;
+			const primaryValue = interval.subtotals[ selectedChart.key ] || 0;
+
+			const secondaryInterval = secondaryDataIntervals[ index ];
+			const secondaryLabel = `${ secondary.label } (${ secondary.range })`;
+			let secondaryLabelDate = secondaryDateMoment.format(
+				'YYYY-MM-DD HH:mm:ss'
+			);
+			let secondaryValue =
+				( secondaryInterval &&
+					secondaryInterval.subtotals[ selectedChart.key ] ) ||
+				0;
+
+			if ( currentInterval === 'day' ) {
+				if (
```

Comment:
It might be worth considering moving the logic to a separate function, as it's getting a bit complex and too nested. 😅  What do you think?

Reply:
Good call! Moved it to utils and also added tests in https://github.com/woocommerce/woocommerce/pull/45874/commits/fd02247f60b49c944e9d32dfee93f53e3f71a822